src/Handler/FileHandler.php

```diff
@@ -42,28 +42,17 @@ protected function download(Composer $composer, IOInterface $io): void
         $target = $this->subpackage->getTargetPath();
         $downloadManager = $composer->getDownloadManager();
 
-        // composer:v2
-        if ($this->isComposerV2()) {
-            $file = '';
-            $promise = $downloadManager->download($this->subpackage, \dirname($target));
-            $promise->then(static function ($res) use (&$file) {
-                $file = $res;
-            });
-            $composer->getLoop()->wait([$promise]);
-            // Look like Composer v2 doesn't care about $target above.
-            // It download the file to "vendor/composer/tmp-[random-file-name]"
-            // We need to move the file to where we want.
-            $this->filesystem->rename($file, $target);
-        }
-        // composer:v1
-        else {
-            // Composer v1 empty the target directory. So we need to create new temporary directory.
-            $tmpDir = \dirname($target).\DIRECTORY_SEPARATOR.uniqid(self::TMP_PREFIX, true);
-            $this->filesystem->ensureDirectoryExists($tmpDir);
-            // Download manager doesn't return the file, so we ask file downloader to do it instead.
-            $file = $downloadManager->getDownloader('file')->download($this->subpackage, $tmpDir);
-            $this->filesystem->rename($file, $target);
-            $this->filesystem->remove($tmpDir);
-        }
+        $file = '';
+        $promise = $downloadManager->download($this->subpackage, \dirname($target));
+        $promise->then(static function ($res) use (&$file) {
+            $file = $res;
+
+            return \React\Promise\resolve($res);
```

Comment:
Do you know which version of composer require this change?

Do we need to update version of `composer-plugin-api` or `composer/composer` to that version?

Reply:
I think we already test after running `composer install --prefer-lowest` and it work fine, so no need to update those packages.

Let me check why I miss this before.

Reply:
I believe it might have been weaker type-checking causing the issue in PHP core. But I will look into this.

Reply:
I tried to remove this line and the tests still passed. So it's not required as I assume. I will resolve as it's a welcome change.

Reply:
I see the problem now on highest dependency:

```
There were 2 errors:

1) LastCall\DownloadsPlugin\Tests\Unit\FileHandlerTest::testInstall
TypeError: Mock_PromiseInterface_5f24811c::then(): Return value must be of type React\Promise\PromiseInterface, null returned

/home/tien/Projects/pact/composer-downloads-plugin/src/Handler/FileHandler.php:47
/home/tien/Projects/pact/composer-downloads-plugin/src/Handler/BaseHandler.php:65
/home/tien/Projects/pact/composer-downloads-plugin/tests/Unit/Handler/BaseHandlerTestCase.php:91

2) LastCall\DownloadsPlugin\Tests\Unit\PharHandlerTest::testInstall
TypeError: Mock_PromiseInterface_5f24811c::then(): Return value must be of type React\Promise\PromiseInterface, null returned

/home/tien/Projects/pact/composer-downloads-plugin/src/Handler/FileHandler.php:47
/home/tien/Projects/pact/composer-downloads-plugin/src/Handler/BaseHandler.php:65
/home/tien/Projects/pact/composer-downloads-plugin/tests/Unit/Handler/BaseHandlerTestCase.php:91


```

So this change + upgrade to composer 2.3 LGTM

Reply:
https://github.com/composer/composer/pull/10547/files#diff-b99c0f08ed0d8f48cb9926a9f015a5cc1fadc6a5c4111c106d788c64264bcb30L187

Which PHP version were you running under? You can see in prior commits it failing based upon this.

Reply:
<img width="955" alt="Screenshot 2024-03-16 at 18 25 14" src="https://github.com/pact-foundation/composer-downloads-plugin/assets/2605791/ed8d972e-6501-43b5-8e41-86dcc30dfb73">